components/k8s-model-server/http-proxy/server.py

```diff
@@ -38,119 +40,184 @@
 
 def _fwrap(f, gf):
-    try:
-        f.set_result(gf.result())
-    except Exception as e:
-        f.set_exception(e)
+  try:
+    f.set_result(gf.result())
+  except Exception as e:
+    f.set_exception(e)
 
 
 def fwrap(gf, ioloop=None):
-    '''
-        Wraps a GRPC result in a future that can be yielded by tornado
-        
-        Usage::
-        
-            @coroutine
-            def my_fn(param):
-                result = yield fwrap(stub.function_name.future(param, timeout))
-        
-    '''
-    f = gen.Future()
-
-    if ioloop is None:
-        ioloop = IOLoop.current()
-
-    gf.add_done_callback(lambda _: ioloop.add_callback(_fwrap, f, gf))
-    return f
+  '''
+  Wraps a GRPC result in a future that can be yielded by tornado
+      
+    Usage::
+      
+      @coroutine
+      def my_fn(param):
+        result = yield fwrap(stub.function_name.future(param, timeout))
+  '''
+  f = gen.Future()
+
+  if ioloop is None:
+    ioloop = IOLoop.current()
+
+  gf.add_done_callback(lambda _: ioloop.add_callback(_fwrap, f, gf))
+  return f
 
 #### END code took from https://github.com/grpc/grpc/wiki/Integration-with-tornado-(python)
 
-def decode_b64_if_needed(value):
-    if isinstance(value, dict):
-        if B64_KEY in value:
-            return base64.b64decode(value[B64_KEY])
-        else:
-            new_value = {}
-            for k, v in value.iteritems():
-                new_value[k] = decode_b64_if_needed(v)
-            return new_value
-    elif isinstance(value, list):
-        new_value = []
-        for v in value:
-            new_value.append(decode_b64_if_needed(v))
-        return new_value
+def decode_b64_if_needed(data):
+  if isinstance(data, list):
+    return [decode_b64_if_needed(val) for val in data]
+  elif isinstance(data, dict):
+    if data.viewkeys() == {"b64"}:
+      return base64.b64decode(data["b64"])
     else:
-        return value
+      return {k: decode_b64_if_needed(v) for k, v in data.iteritems()}
+  else:
+    return data
+
+def get_signature_map(model_server_stub, model_name):
+  request = get_model_metadata_pb2.GetModelMetadataRequest()
+  request.model_spec.name = model_name
+  request.metadata_field.append("signature_def")
+  try:
+    response = model_server_stub.GetModelMetadata(request, MODEL_SERVER_METADATA_TIMEOUT_SEC)
+  except grpc.RpcError as rpc_error:
+    logging.exception("GetModelMetadata call to model server failed with code "
+                      "%s and message %s", rpc_error.code(),
+                      rpc_error.details())
+    return None
+
+  signature_def_map_proto = get_model_metadata_pb2.SignatureDefMap()
+  response.metadata["signature_def"].Unpack(signature_def_map_proto)
+  signature_def_map = signature_def_map_proto.signature_def
+  if not signature_def_map:
+    logging.error("Graph has no signatures.")
+
+  # Delete incomplete signatures without input dtypes.
+  invalid_signatures = []
+  for signature_name in signature_def_map:
```

Comment:
would this fail tensorflow/serving already ?

Reply:
Not sure, but here we will print out the explicit warning message, I think it will be more informative.